File: sdk/go/policy_learner.go
package alpgo

import (
	"strings"
)

type PolicyContext struct {
	ContextID   string            `json:"context_id"`
	Environment map[string]string `json:"environment"`
	Tags        []string          `json:"tags"`
}

type PolicyLearner struct {
	history []PolicyContext
}

func NewPolicyLearner() *PolicyLearner {
	return &PolicyLearner{history: []PolicyContext{}}
}

func (p *PolicyLearner) Learn(ctx PolicyContext) {
	p.history = append(p.history, ctx)
}

func (p *PolicyLearner) Suggest(env map[string]string) []string {
	var candidates []string
	for _, ctx := range p.history {
		if matchEnvironment(env, ctx.Environment) {
			candidates = append(candidates, strings.Join(ctx.Tags, ","))
		}
	}
	return dedupe(candidates)
}

func (p *PolicyLearner) History() []PolicyContext {
	out := make([]PolicyContext, len(p.history))
	copy(out, p.history)
	return out
}

func matchEnvironment(a, b map[string]string) bool {
	if len(a) == 0 || len(b) == 0 {
		return false
	}
	matches := 0
	for k, v := range a {
		if b[k] == v {
			matches++
		}
	}
	return matches > 0
}

func dedupe(in []string) []string {
	seen := map[string]bool{}
	var out []string
	for _, s := range in {
		if !seen[s] {
			seen[s] = true
			out = append(out, s)
		}
	}
	return out
}

File: sdk/rust/src/policy_learner.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PolicyContext {
    pub context_id: String,
    pub environment: HashMap<String, String>,
    pub tags: Vec<String>,
}

pub struct PolicyLearner {
    history: Vec<PolicyContext>,
}

impl PolicyLearner {
    pub fn new() -> Self {
        Self {
            history: Vec::new(),
        }
    }

    pub fn learn(&mut self, ctx: PolicyContext) {
        self.history.push(ctx);
    }

    pub fn suggest(&self, env: &HashMap<String, String>) -> Vec<String> {
        let mut candidates = Vec::new();
        for ctx in &self.history {
            if Self::match_environment(env, &ctx.environment) {
                candidates.push(ctx.tags.join(","));
            }
        }
        Self::dedupe(candidates)
    }

    pub fn history(&self) -> Vec<PolicyContext> {
        self.history.clone()
    }

    fn match_environment(a: &HashMap<String, String>, b: &HashMap<String, String>) -> bool {
        if a.is_empty() || b.is_empty() {
            return false;
        }
        let mut matches = 0;
        for (k, v) in a {
            if b.get(k) == Some(v) {
                matches += 1;
            }
        }
        matches > 0
    }

    fn dedupe(mut input: Vec<String>) -> Vec<String> {
        input.sort();
        input.dedup();
        input
    }
}
